### PR TITLE
[HOSTEDCP-1041] Defaulting webhook for self managed HCP

### PR DESCRIPTION
**Pull request by davidvossel**
Previously, the hostedCluster.Spec.Release.Image was required and the hcp or hypershift cli defaulted that value client side. This posed several issues though.

1. The defaulting did not take into consideration what the max and min version the deployed hypershift operator supports, which will eventually lead to the client defaulting unsupported versions.
2. This logic was all client side, which meant that it couldn't be used in a gitops flow or reused by the console web UI.

The new logic performs defaulting the release image on the backend using an **optional** mutating webhook. The logic takes into account the latest/min supported versions and picks the most recent version that falls within those constraints.

A webhook was chosen over performing defaults in the controller's reconcile loop due to the following reasons. 
* The Webhook requires no api changes (such as making the release image optional)
* Error discovery occurs immediately at creation time (http Post response) rather than after admission via condition

### CODE REVIEW

File: api/v1beta1/hostedcluster_types.go [hunk 1]
```diff
@@ -1659,7 +1659,7 @@ type ManagedEtcdStorageSpec struct {
 	//
 	// +optional
 	// +immutable
-	RestoreSnapshotURL []string `json:"restoreSnapshotURL"`
```
Comment (sjenning) [inline api/v1beta1/hostedcluster_types.go:1662]: Why this change?

File: cmd/cluster/core/create.go [hunk 2]
```diff
@@ -440,7 +432,11 @@ func CreateCluster(ctx context.Context, opts *CreateOptions, platformSpecificApp
 func defaultNetworkType(ctx context.Context, opts *CreateOptions, releaseProvider releaseinfo.Provider, readFile func(string) ([]byte, error)) error {
 	if opts.NetworkType != "" {
 		return nil
+	} else if opts.ReleaseImage == "" {
```
Comment (sjenning) [inline cmd/cluster/core/create.go:435]: I think this change is causing the unit test failure

File: support/supportedversion/version.go [hunk 3]
```diff
@@ -76,3 +80,43 @@ func IsValidReleaseVersion(version, currentVersion, latestVersionSupported, minS
...
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	err = json.Unmarshal(body, &version)
+	if err != nil {
+		return "", err
+	}
+	return version.PullSpec, nil
+
```
Comment (sjenning) [inline support/supportedversion/version.go:121]: nit: extra line

File: hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go [hunk 4]
```diff
@@ -1,22 +1,85 @@
...
+
+func (defaulter *nodePoolDefaulter) Default(ctx context.Context, obj runtime.Object) error {
+	np, ok := obj.(*hyperv1.NodePool)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a NodePool but got a %T", obj))
+	}
+
+	if np.Spec.Release.Image != "" {
+		return nil
+	} else if np.Spec.ClusterName == "" {
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go:50]: Why is the defaulter doing throwing this validation error? the .ClusterName is already required an immutable

File: support/supportedversion/version.go [hunk 5]
```diff
@@ -76,3 +80,43 @@ func IsValidReleaseVersion(version, currentVersion, latestVersionSupported, minS
...
 }
+
+type ocpVersion struct {
+	Name        string `json:"name"`
+	PullSpec    string `json:"pullSpec"`
+	DownloadURL string `json:"downloadURL"`
+}
+
+func LookupLatestSupportedRelease(ctx context.Context) (string, error) {
+	prefix := "https://multi.ocp.releases.ci.openshift.org/api/v1/releasestream/4-stable-multi/latest"
```
Comment (enxebre) [inline support/supportedversion/version.go:91]: Did we make a decision of defaulting to multiarch? This has an impact particularly for disconnected environments https://issues.redhat.com/browse/MIXEDARCH-292 / https://issues.redhat.com/browse/MIXEDARCH-255

At minimum I thin we should document and communicate this clearly in docs?

File: support/supportedversion/version.go [hunk 6]
```diff
@@ -76,3 +80,43 @@ func IsValidReleaseVersion(version, currentVersion, latestVersionSupported, minS
 
 	return nil
 }
+
+type ocpVersion struct {
+	Name        string `json:"name"`
+	PullSpec    string `json:"pullSpec"`
+	DownloadURL string `json:"downloadURL"`
+}
+
```
Comment (enxebre) [inline support/supportedversion/version.go:89]: would you mind document this func? e.g `// LookupLatestSupportedRelease picks the latest multi-arch image supported by this Hypershift Operator...`

**Comment by enxebre** [comment]
it'd be nice to exercise this in the kubevirt job

File: api/v1beta1/hostedcluster_types.go (hunk 1 above)
Comment (davidvossel) [reply api/v1beta1/hostedcluster_types.go:1662]: i know... I thought so too. but it actually matters in a kind of dumb way (that I only half understand)

The issue is with how the decoding/encoding of the object works in the webhook. If we don't add `omitempty` here, then the decode/encode logic results in the mutation webhook trying to patch RestoreSnapshotURL to `Null`, then we get an API error during the validation phase about trying to set a `[]string` to a Null value... or something along those lines.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go (hunk 4 above)
Comment (davidvossel) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go:50]: Two reasons

1. I don't think ClusterName required to be a non empty string. I believe it is only required to be immutable. this is the validation rule for cluster name `+kubebuilder:validation:XValidation:rule="self == oldSelf", message="ClusterName is immutable"`

2. The mutation webhook runs before the validation, which is why we can default the empty release image even though it is required to not be empty via this validation rule `+kubebuilder:validation:Pattern=^(\w+\S+)$`


So, i'm kind of in a weird spot with the cluster name here.

File: support/supportedversion/version.go (hunk 5 above)
Comment (davidvossel) [reply support/supportedversion/version.go:91]: The cli defaulted to multi, so i brought that over to the webhook. https://github.com/openshift/hypershift/blob/1fc3c32155a1b12fc964812ea310aefd313e07ca/cmd/version/version.go#L22 

If someone is using disconnected, I doubt they'd be using the release image defaulting.

**Comment by davidvossel** [comment]
> it'd be nice to exercise this in the kubevirt job
 
yep, I think it will take a few steps for us to get to that point though. We'll need to enable the webhook during install (which involves messing with openshift/release ci code), then we can add a e2e test case that ensures the HC/NP defaults the release image during creation when kubevirt platform is in use.

File: cmd/cluster/core/create.go (hunk 2 above)
Comment (davidvossel) [reply cmd/cluster/core/create.go:435]: ah, yep, fixed

File: hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go [hunk 7]
```diff
@@ -1,22 +1,85 @@
...
+	hc := &hyperv1.HostedCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      np.Spec.ClusterName,
+			Namespace: np.Namespace,
+		},
+	}
+
+	err := defaulter.client.Get(ctx, client.ObjectKeyFromObject(hc), hc)
+	if err != nil {
+		return fmt.Errorf("Error retrieving HostedCluster named [%s], %v", np.Spec.ClusterName, err)
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go:63]: Cap letter here makes verify fail.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go (hunk 4 above)
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go:50]: you're right though this ensures presence but it would accept an explicitly set zero value for the string, and as you mentioned if we enforced != "" validation at the crd schema it would run after this, so I'm fine either way.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go (hunk 7 above)
Comment (davidvossel) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go]: oops, thanks, fixed
